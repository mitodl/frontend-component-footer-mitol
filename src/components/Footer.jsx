import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { ensureConfig } from '@edx/frontend-platform';
import { AppContext } from '@edx/frontend-platform/react';

import messages from './Footer.messages';
import LanguageSelector from './LanguageSelector';

ensureConfig([
  'LMS_BASE_URL',
  'LOGO_TRADEMARK_URL',
], 'Footer component');

const EVENT_NAMES = {
  FOOTER_LINK: 'edx.bi.footer.link',
};

class SiteFooter extends React.Component {
  constructor(props) {
    super(props);
    this.externalLinkClickHandler = this.externalLinkClickHandler.bind(this);
  }

  externalLinkClickHandler(event) {
    const label = event.currentTarget.getAttribute('href');
    const eventName = EVENT_NAMES.FOOTER_LINK;
    const properties = {
      category: 'outbound_link',
      label,
    };
    sendTrackEvent(eventName, properties);
  }

  renderLinkIfExists(value, text) {
    return value && <li><a href={value}>{text}</a></li>;
  }

  render() {
    const {
      supportedLanguages,
      onLanguageSelected,
      logo,
      intl,
    } = this.props;
    const showLanguageSelector = supportedLanguages.length > 0 && onLanguageSelected;
    const { config } = this.context;
    return (
      <footer
        role="contentinfo"
        className="footer d-flex border-top py-3 px-4"
      >
        <div className="container-fluid d-flex">
          { process.env.SHOW_LOGO
          && (
          <a
            className="d-block"
            href={process.env.SITE_URL}
            aria-label={intl.formatMessage(messages['footer.logo.ariaLabel'])}
          >
            <img
              style={{ maxHeight: 45 }}
              src={logo || config.LOGO_TRADEMARK_URL}
              alt={process.env.LOGO_ALT_TEXT || intl.formatMessage(messages['footer.logo.altText'])}
            />
          </a>
          )}
          <div className="copyright-col">
            {process.env.TRADEMARK_TEXT
            && (
            <div className="text-gray-500 small">
                {process.env.TRADEMARK_TEXT}
            </div>
            )}
            <div>
              <ul className="footer-sub-nav">
                {this.renderLinkIfExists(process.env.ABOUT_US_URL, 'About Us')}
                {this.renderLinkIfExists(process.env.TERMS_OF_SERVICE_URL, 'Terms of Service')}
                {this.renderLinkIfExists(process.env.PRIVACY_POLICY_URL, 'Privacy Policy')}
                {this.renderLinkIfExists(process.env.HONOR_CODE_URL, 'Honor Code')}
                {this.renderLinkIfExists(process.env.Contact, 'Contact')}
                {this.renderLinkIfExists(process.env.SUPPORT_CENTER_URL, process.env.SUPPORT_CENTER_TEXT || 'FAQ & Help')}
              </ul>
            </div>
          </div>
          <div className="flex-grow-1" />
          {showLanguageSelector && (
            <LanguageSelector
              options={supportedLanguages}
              onSubmit={onLanguageSelected}
            />
          )}
        </div>
      </footer>
    );
  }
}

SiteFooter.contextType = AppContext;

SiteFooter.propTypes = {
  intl: intlShape.isRequired,
  logo: PropTypes.string,
  onLanguageSelected: PropTypes.func,
  supportedLanguages: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
  })),
};

SiteFooter.defaultProps = {
  logo: undefined,
  onLanguageSelected: undefined,
  supportedLanguages: [],
};

export default injectIntl(SiteFooter);
export { EVENT_NAMES };
