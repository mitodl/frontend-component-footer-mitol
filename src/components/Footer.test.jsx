/* eslint-disable react/prop-types */
import React, { useMemo } from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent, screen } from '@testing-library/react';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import { AppContext } from '@edx/frontend-platform/react';

import Footer from './Footer';

const FooterWithContext = ({ locale = 'es' }) => {
  const contextValue = useMemo(() => ({
    authenticatedUser: null,
    config: {
      LOGO_TRADEMARK_URL: process.env.LOGO_TRADEMARK_URL,
      LMS_BASE_URL: process.env.LMS_BASE_URL,
      ABOUT_US_URL: process.env.ABOUT_US_URL,
      ACCESSIBILITY_URL: process.env.ACCESSIBILITY_URL,
      CONTACT_URL: process.env.CONTACT_URL,
      HONOR_CODE_URL: process.env.HONOR_CODE_URL,
      FOOTER_LOGO_ALT_TEXT: process.env.FOOTER_LOGO_ALT_TEXT,
      PRIVACY_POLICY_URL: process.env.PRIVACY_POLICY_URL,
      SHOW_FOOTER_LOGO: process.env.SHOW_FOOTER_LOGO,
      SUPPORT_CENTER_TEXT: process.env.SUPPORT_CENTER_TEXT,
      SUPPORT_CENTER_URL: process.env.SUPPORT_CENTER_URL,
      TERMS_OF_SERVICE_URL: process.env.TERMS_OF_SERVICE_URL,
      TRADEMARK_TEXT: process.env.TRADEMARK_TEXT,
    },
  }), []);

  return (
    <IntlProvider locale={locale}>
      <AppContext.Provider
        value={contextValue}
      >
        <Footer />
      </AppContext.Provider>
    </IntlProvider>
  );
};

const FooterWithLanguageSelector = ({ languageSelected = () => {} }) => {
  const contextValue = useMemo(() => ({
    authenticatedUser: null,
    config: {
      LOGO_TRADEMARK_URL: process.env.LOGO_TRADEMARK_URL,
      LMS_BASE_URL: process.env.LMS_BASE_URL,
      ABOUT_US_URL: process.env.ABOUT_US_URL,
      ACCESSIBILITY_URL: process.env.ACCESSIBILITY_URL,
      CONTACT_URL: process.env.CONTACT_URL,
      HONOR_CODE_URL: process.env.HONOR_CODE_URL,
      FOOTER_LOGO_ALT_TEXT: process.env.FOOTER_LOGO_ALT_TEXT,
      PRIVACY_POLICY_URL: process.env.PRIVACY_POLICY_URL,
      SHOW_FOOTER_LOGO: process.env.SHOW_FOOTER_LOGO,
      SUPPORT_CENTER_TEXT: process.env.SUPPORT_CENTER_TEXT,
      SUPPORT_CENTER_URL: process.env.SUPPORT_CENTER_URL,
      TERMS_OF_SERVICE_URL: process.env.TERMS_OF_SERVICE_URL,
      TRADEMARK_TEXT: process.env.TRADEMARK_TEXT,
    },
  }), []);

  return (
    <IntlProvider locale="en">
      <AppContext.Provider
        value={contextValue}
      >
        <Footer
          onLanguageSelected={languageSelected}
          supportedLanguages={[
            { label: 'English', value: 'en' },
            { label: 'EspaÃ±ol', value: 'es' },
          ]}
        />
      </AppContext.Provider>
    </IntlProvider>
  );
};

describe('<Footer />', () => {
  describe('renders correctly', () => {
    it('renders without a language selector', () => {
      const tree = renderer
        .create(<FooterWithContext locale="en" />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders without a language selector in es', () => {
      const tree = renderer
        .create(<FooterWithContext locale="es" />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with a language selector', () => {
      const tree = renderer
        .create(<FooterWithLanguageSelector />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('handles language switching', () => {
    it('calls onLanguageSelected prop when a language is changed', () => {
      const mockHandleLanguageSelected = jest.fn();
      render(<FooterWithLanguageSelector languageSelected={mockHandleLanguageSelected} />);

      fireEvent.submit(screen.getByTestId('site-footer-submit-btn'), {
        target: {
          elements: {
            'site-footer-language-select': {
              value: 'es',
            },
          },
        },
      });

      expect(mockHandleLanguageSelected).toHaveBeenCalledWith('es');
    });
  });
});
