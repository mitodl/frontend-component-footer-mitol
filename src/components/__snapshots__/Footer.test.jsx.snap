// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> renders correctly renders with a language selector 1`] = `
<footer
  className="footer d-flex border-top py-3 px-4"
  role="contentinfo"
>
  <div
    className="container-fluid d-flex"
  >
    <a
      aria-label="edX Home"
      className="d-block"
      href="http://localhost:18000/"
    >
      <img
        alt="alt text"
        src="https://edx-cdn.org/v3/default/logo-trademark.svg"
        style={
          Object {
            "maxHeight": 45,
          }
        }
      />
    </a>
    <div
      className="copyright-col"
    >
      <div
        className="text-gray-500 small"
      >
        Trade Mark text
      </div>
      <div>
        <ul
          className="footer-sub-nav"
        >
          <li>
            <a
              href="http://localhost:18000/about"
            >
              About Us
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/tos_and_honor"
            >
              Terms of Service
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/privacy"
            >
              Privacy Policy
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/tos_and_honor"
            >
              Honor Code
            </a>
          </li>
          <li>
            <a
              href="https://accessibility.mit.edu/"
            >
              Accessibility
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/about"
            >
              SUPPORT CENTER
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="flex-grow-1"
    />
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          className="d-inline-block m-0"
          htmlFor="site-footer-language-select"
        >
          Choose Language
        </label>
        <select
          className="form-control-sm mx-2"
          defaultValue="en"
          id="site-footer-language-select"
          name="site-footer-language-select"
        >
          <option
            value="en"
          >
            English
          </option>
          <option
            value="es"
          >
            Espa√±ol
          </option>
        </select>
        <button
          className="btn btn-outline-primary btn-sm"
          type="submit"
        >
          Apply
        </button>
      </div>
    </form>
  </div>
</footer>
`;

exports[`<Footer /> renders correctly renders without a language selector 1`] = `
<footer
  className="footer d-flex border-top py-3 px-4"
  role="contentinfo"
>
  <div
    className="container-fluid d-flex"
  >
    <a
      aria-label="edX Home"
      className="d-block"
      href="http://localhost:18000/"
    >
      <img
        alt="alt text"
        src="https://edx-cdn.org/v3/default/logo-trademark.svg"
        style={
          Object {
            "maxHeight": 45,
          }
        }
      />
    </a>
    <div
      className="copyright-col"
    >
      <div
        className="text-gray-500 small"
      >
        Trade Mark text
      </div>
      <div>
        <ul
          className="footer-sub-nav"
        >
          <li>
            <a
              href="http://localhost:18000/about"
            >
              About Us
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/tos_and_honor"
            >
              Terms of Service
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/privacy"
            >
              Privacy Policy
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/tos_and_honor"
            >
              Honor Code
            </a>
          </li>
          <li>
            <a
              href="https://accessibility.mit.edu/"
            >
              Accessibility
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/about"
            >
              SUPPORT CENTER
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="flex-grow-1"
    />
  </div>
</footer>
`;

exports[`<Footer /> renders correctly renders without a language selector in es 1`] = `
<footer
  className="footer d-flex border-top py-3 px-4"
  role="contentinfo"
>
  <div
    className="container-fluid d-flex"
  >
    <a
      aria-label="edX Home"
      className="d-block"
      href="http://localhost:18000/"
    >
      <img
        alt="alt text"
        src="https://edx-cdn.org/v3/default/logo-trademark.svg"
        style={
          Object {
            "maxHeight": 45,
          }
        }
      />
    </a>
    <div
      className="copyright-col"
    >
      <div
        className="text-gray-500 small"
      >
        Trade Mark text
      </div>
      <div>
        <ul
          className="footer-sub-nav"
        >
          <li>
            <a
              href="http://localhost:18000/about"
            >
              About Us
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/tos_and_honor"
            >
              Terms of Service
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/privacy"
            >
              Privacy Policy
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/tos_and_honor"
            >
              Honor Code
            </a>
          </li>
          <li>
            <a
              href="https://accessibility.mit.edu/"
            >
              Accessibility
            </a>
          </li>
          <li>
            <a
              href="http://localhost:18000/about"
            >
              SUPPORT CENTER
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="flex-grow-1"
    />
  </div>
</footer>
`;
